### ISSUE TITLE
CUDA build using PyPI releases not working

### ISSUE THREAD

**Issue by jklaise** [shap/shap#1834]
Following #1650 I was able to build `shap` with the cuda extension from source by just running `python setup.py bdist_wheel` (with cuda 10.1, no environment variables set). However, instead of going through the process of cloning the repo, building the wheel manually and then installing it, I wanted to make `pip` do it automatically. Forcing a wheel build and logging errors with the following command:

```
pip install shap --no-cache-dir -vvv
```

results in the following error and the cuda extension not built as a result:

```
Running setup.py (path:/tmp/pip-install-__dq7b2a/shap/setup.py) egg_info for package shap
    Created temporary directory: /tmp/pip-pip-egg-info-k3cm5emn
    Running command python setup.py egg_info
    gcc: error: shap/cext/_cext_gpu.cu: No such file or directory
    gcc: warning: [34m~@~X-x c++[34m~@~Y after last input file has no effect
    gcc: fatal error: no input files
    compilation terminated.
    NVCC ==>  /usr/local/cuda-10.1/bin/nvcc
    Compiling cuda extension, calling nvcc with arguments:
    ['/usr/local/cuda-10.1/bin/nvcc', 'shap/cext/_cext_gpu.cu', '-lib', '-o', 'build/lib_cext_gpu.a', '-Xcompiler', '-fPIC', '-I/home/janis/miniconda3/envs/tf2/include/python3.8', '--std', 'c++14', '--expt-extended-lambda', '--expt-relaxed-constexpr', '-arch=sm_60', '-gencode=arch=compute_70,code=sm_70', '-gencode=arch=compute_75,code=sm_75', '-gencode=arch=compute_75,code=compute_75']
    Error building cuda module: CalledProcessError(1, ['/usr/local/cuda-10.1/bin/nvcc', 'shap/cext/_cext_gpu.cu', '-lib', '-o', 'build/lib_cext_gpu.a', '-Xcompiler', '-fPIC', '-I/home/janis/miniconda3/envs/tf2/include/python3.8', '--std', 'c++14', '--expt-extended-lambda', '--expt-relaxed-constexpr', '-arch=sm_60', '-gencode=arch=compute_70,code=sm_70', '-gencode=arch=compute_75,code=sm_75', '-gencode=arch=compute_75,code=compute_75'])
    WARNING: Could not compile cuda extensions
```

As you can see the build process can't find the extension. I have a suspicion this may be because the extensions are not part of the `MANIFEST.in` which is now outdated after extensions where moved under `shap/cext`.

**Comment by jklaise** [comment]
Actually, I'm not sure it's an issue with `MANIFEST.in` as I was able to compile and install by just pointing to the git repo as the source:

```
pip install git+git://github.com/slundberg/shap@v0.38.1#egg=shap -vvv
```

**Comment by jklaise** [comment]
Still the same issue with the new `0.39` version. I looked inside the `shap-0.39.0.tar.gz` sdist downloaded from PyPI and it is missing files in the `cext` directory:

```
(shap-test) janis@lambda-quad:~/GIT/packages/shap-0.39.0/shap/cext$ ls                                                                                                                                              
_cext.cc  tree_shap.h                                                                                                                                                                                               
```
So when creating the `sdist` somehow the other extensions are missing. When I run `python setup.py sdist` the created `sdist` does contain all files under `shap/cext` so not sure how it is possible that these are missing from the official `sdist`.

@slundberg how is the `sdist` generated and is there any output info to see why the extra files are missing?

**Comment by jklaise** [comment]
Confirming that the issue persists in the latest `0.41` version.

**Comment by jklaise** [comment]
## Update for the current state of this issue. 

tl;dr:
 - There is no GPU support for the `sdist` published on PyPI as the cuda sources are missing
 - There appears to be no GPU support for the pre-build wheels either (not surprising as need a machine with GPU and `nvcc`
 - We can make sure GPU support is enabled by including the cuda sources in the `sdist` published on PyPI (otherwise people need to clone the repo to do a GPU-enabled build). This can be done by including a `MANIFEST.in` file which the `build` tool will automatically pick-up
 - GPU support for pre-built wheels might be trickier as requires `nvcc` and a GPU on the build machine

---

Since `0.40` `shap` has shipped with an extensive selection of wheels, including `manylinux` which ship with compiled `cuda` extensions. Thus, most `pip install` commands will pick-up one of these wheels instead of trying to build from `sdist`.

However, the published `sdist` is still missing the source of the cuda extension, so if someone attempted to build their own wheel with cuda support, it would fail (a workaround here, as before, would be to clone the repository and make a local build instead of relying on the `sdist` published on PyPI which is missing the `cuda` sources. To force installing from source via PyPI one can use the following command (run on a system with `nvcc`):

```bash
pip install shap --no-cache-dir --no-binary shap -vvv
```
The package will build, but without `cuda` support (assuming you're running on a machine with `nvcc` and a GPU. To check that GPU support is missing, after the installation do (run using `numpy==1.22.4` due to newer versions throwing an error as `shap` uses deprecated `np.bool`):

```python
from shap.explainers import GPUTree
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

X, y = make_classification()
model = RandomForestClassifier().fit(X, y)
exp = GPUTree(model)

exp.shap_values(X[:3])
```
```
cuda extension was not built during install!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/janis/miniconda3/envs/py310/lib/python3.10/site-packages/shap/explainers/_gpu_tree.py", line 105, in shap_values
    assert_import("cext_gpu")
  File "/home/janis/miniconda3/envs/py310/lib/python3.10/site-packages/shap/utils/_general.py", line 25, in assert_import
    raise e
  File "/home/janis/miniconda3/envs/py310/lib/python3.10/site-packages/shap/explainers/_gpu_tree.py", line 6, in <module>
    from .. import _cext_gpu
ImportError: cannot import name '_cext_gpu' from partially initialized module 'shap' (most likely due to a circular import) (/home/janis/miniconda3/envs/py310/lib/python3.10/site-packages/shap/__init__.py)
```

**NB:** You can trigger the same error by installing one of the `shap` built wheels, which suggests that GPU support is also missing from the pre-built wheels:
```
pip install shap --no-cache-dir -vvv
```

## Partial remedy
We can enable GPU support in the `sdist` by simply packaging the cuda sources into it, then anyone on a machine with `nvcc` and a GPU can build from source by pulling the `sdist` from PyPI. This would still require people to explicitly pass `--no-binary shap` flag to `pip` since the pre-built wheels lack GPU support. This requires an update to `setup.py` (or perhaps `MANIFEST.in` to ship the source files in the `sdist` - currently the sources are only included conditionally on whether the `with_binary` and `with_cuda` flags in `setup.py` are satisfied (depends on the build system). Contents of `MANIFEST.in`:

```
include shap/cext/*
```
